modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/get_portlet_namespace.ts

```diff
@@ -14,11 +14,8 @@
 
 /**
  * Returns the portlet namespace with underscores prepended and appended to it
- * @param {!string} portletId The portlet ID to add underscores to
- * @return {string} Portlet namespace with underscores prepended and appended
- * @review
  */
-export default function getPortletNamespace(portletId) {
+export default function getPortletNamespace(portletId: string) {
```

Comment:
Technically, this takes `any`... so... runtime type guards, keep'em or drop'em? Should we be stricter with types like this, or better type as our current usage and then clean it up?

Reply:
> runtime type guards, keep'em or drop'em?

I would say you still need them in case someone calls you from plain JavaScript...

However, if we decide to do runtime type checks, we should create a small utility for that, so that code is cleaner and less error prone. Something like:

```
Ensure.isString(portletId);
```

That we put at the top of methods.

Reply:
> I would say you still need them in case someone calls you from plain JavaScript...

The thing that worries me here is that we're throwing `TypeError` exceptions on a bad call. It's not like we're gracefully putting the system in a place where it can recover, but rather crashing it to the ground at runtime. 

This case is particularly tricky because the main difference is that if a developer calls this like `getPortletNamespace(3)` it will get:

```
// With runtime type guards
Uncaught TypeError: portletId must be a string

// Without
_3_
```

At runtime, the uncaught error seems more harmful, although the silent valid return can let issues go unnoticed to the developer for a while.

The more common case I guess would be something like `addParams(3, 'http://www.liferay.com')`

```
// With runtime type guards
Uncaught TypeError: Parameter params must be an object or string

// Without
Uncaught TypeError: params.trim is not a function
```

Now, I'm not convinced one way or another, so just trying to keep the conversation going here. It's not like we have so many of these anyways, but maybe we could keep the runtime type checks for utilities that get exported globally (ie: accessible through `Liferay.foo`) but not for those that need you to `import {foo} from 'frontend-js-web`', assuming the latter will be more easily moved into TS

Reply:
I agree that runtime failures are nasty, but people can avoid them simply by using a linter or TS at build time. If they don't want to do that, then they are exposed to more frequent runtime failures, because that's one of the reasons why typed languages were invented :shrug: .

Also, not throwing is much worse because you are deferring the failure to -basically- a random point in the future or, even worse, let it fail silently in ways that are very hard to reason about.

Say the developer writes `addParams(3, 'http://www.liferay.com')` and, instead of failing with `Uncaught TypeError: params.trim is not a function`, the code simply ignores the second parameter (because it uses an `Object.keys` or something like that, instead of `trim`).

Then, the method invocation would completely ignore the `'http://www.liferay.com'` but the developer would be quite puzzled because all he sees is that he's passing the option (correctly, he thinks) but the method is not honoring it. How long will it take to realize what is the error?

And in production, this will lead to some kind of functional failure which users will suffer too, so, is that better than failing fast with an exception? I doubt it...

Reply:
Think about security too. If you use parameters as a type, but the developers provide other types, it can lead to subtle vulnerabilites (as with user input). In an API, any parameter should be considered like user input, IMO, since you don't have any control over it.

Reply:
> And in production, this will lead to some kind of functional failure which users will suffer too, so, is that better than failing fast with an exception? I doubt it...

If you're in production, you probably haven't failed fast enough, have you? 😝 

I get all the pros/cons of the different approaches, so I think the point I'm trying to make is... what should be our default and how consistent we want to be about it? Right now, we only do runtime type checks in a handful of places but most others don't have any guarantees. As an example, [`fetch`](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/fetch.es.js), which is a widely used utility doesn't do any type checking.

Should we:
- Drop all runtime type checks for typed APIs
- Keep (and add new) runtime type checks based on some rule of thumb. If so, which APIs should have checks and which ones shouldn't?
- Add runtime type checks everywhere

(I know this is a bit of a reductio ad absurdum because we probably don't want to do the last thing... or do we? 😉 )

Reply:
>  With TS we can still get runtime errors

_[Elm](https://elm-lang.org/) has entered the chat_

Reply:
:joy:

Reply:
>  I really don't want to make a distinction between people who choose TS and those who choose JS

It's not a distinction between TS and JS but between having a build with type checking or no build at all. Wether that type checking build is implemented with TS, eslint, or any other tool is an implementation detail.

Regarding runtime checks, I'm a big fan of super-robust public APIs for typed languages as well as dynamic ones, so I would make them (and probably design some way to make them automatic, not hand made), as I said above :shrug:

Reply:
> It's not a distinction between TS and JS

I was referring to this part :wink:

> ... people using plain JS are on their own (and should suffer, at least a bit, for choosing not to use types).

But, I understand what you said about runtime checks

Reply:
> (and should suffer, at least a bit, for choosing not to use types)

<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia1.giphy.com%2Fmedia%2FdeaabWAgiawzyiMNZS%2Fgiphy.gif&token=kill" data-gitmeme-token="kill"><img src="https://media1.giphy.com/media/deaabWAgiawzyiMNZS/giphy.gif" title="Created by gitme.me with /kill"/></a>